Lib/inspect.py

```diff
@@ -792,25 +792,42 @@ def findsource(object):
         return lines, 0
 
     if isclass(object):
-        name = object.__name__
-        pat = re.compile(r'^(\s*)class\s*' + name + r'\b')
-        # make some effort to find the best matching class definition:
-        # use the one with the least indentation, which is the one
-        # that's most probably not inside a function definition.
-        candidates = []
-        for i in range(len(lines)):
-            match = pat.match(lines[i])
-            if match:
-                # if it's at toplevel, it's already the best one
-                if lines[i][0] == 'c':
-                    return lines, i
-                # else add whitespace to candidate list
-                candidates.append((match.group(1), i))
-        if candidates:
-            # this will sort by whitespace, and by line number,
-            # less whitespace first
-            candidates.sort()
-            return lines, candidates[0][1]
+        import ast
+
+        class ClassVisitor(ast.NodeVisitor):
+
+            def __init__(self, qualname):
```

Comment:
Since the class is not used outside of `findsource()` you can simplify the code by removing `__init__`. Inside the visitor you can just use local variables from the outer scope. This will save several lines. Add `nonlocal line_number` because it is set inside the visitor.

Reply:
I tried removing the constructor and using local variables which removes use of `self` as in `self.stack`. But since stack is changed in three functions now I have to declare `nonlocal` on three and make other changes. After conversion I think using class variables is more self-contained and I have kept them. Let me know if you need to change them to use local variables.

Reply:
No, you don't have to use `nonlocal` with a stack, because you don't assign to this variable in these methods. Only `line_number` needs it, at only in a single place. I prefer more compact code, but it is up to you using the construct and instance attributes.